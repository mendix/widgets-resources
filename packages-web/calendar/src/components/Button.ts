import { ReactNode, createElement, FunctionComponent } from "react";
import classNames from "classnames";

export interface ButtonProps {
    renderMode: "button" | "link";
    title?: string;
    className?: string;
    buttonStyle?: ButtonStyle;
    icon?: string;
    iconPosition?: "left" | "right";
    active: boolean;
    caption: ReactNode;
    onClick: () => void;
}

export type ButtonStyle = "default" | "primary" | "success" | "info" | "inverse" | "warning" | "danger";

export const addIcon = (icon?: string, iconPosition?: "left" | "right", content?: ReactNode): ReactNode => {
    if (icon) {
        if (iconPosition === "right") {
            return [content, content ? " " : null, createElement("span", { className: icon })];
        } else {
            return [createElement("span", { className: icon }), content ? " " : null, content];
        }
    } else {
        return content;
    }
};

export const ToolbarButton: FunctionComponent<ButtonProps> = props =>
    props.renderMode === "button" ? Button(props) : Link(props);

ToolbarButton.defaultProps = { buttonStyle: "default", renderMode: "button", iconPosition: "left" };

export const Button: FunctionComponent<ButtonProps> = ({
    title,
    className,
    caption,
    onClick,
    buttonStyle,
    active,
    icon,
    iconPosition
}) =>
    createElement(
        "button",
        {
            className: classNames("btn", `btn-${buttonStyle}`, className, { active }),
            title,
            onClick
        },
        addIcon(icon, iconPosition, caption)
    );

const Link: FunctionComponent<ButtonProps> = ({ title, className, caption, onClick, active, icon, iconPosition }) =>
    createElement(
        "span",
        {
            className: classNames("mx-link", className, { active }),
            tabindex: "0",
            title,
            onClick
        },
        addIcon(icon, iconPosition, createElement("a", { tabindex: "-1" }, caption))
    );
