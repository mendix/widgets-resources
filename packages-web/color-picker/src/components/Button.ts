import { SFC, createElement } from "react";
import classNames from "classnames";

import { Mode } from "./ColorPicker";

export interface ButtonProps {
    className?: string;
    disabled: boolean;
    color?: string;
    mode: Mode;
    onClick?: () => void;
    tabIndex?: number;
}

export const Button: SFC<ButtonProps> = props =>
    createElement(
        "button",
        {
            className: classNames("btn", {
                "widget-color-picker-input": props.mode === "input",
                hidden: props.mode === "inline",
                disabled: props.disabled
            }),
            onClick: props.onClick,
            tabIndex: props.tabIndex
        },
        createElement("div", { className: props.className, style: { background: props.color } })
    );

Button.displayName = "Button";
