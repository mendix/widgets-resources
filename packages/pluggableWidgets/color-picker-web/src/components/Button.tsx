import { createElement, ReactElement } from "react";
import classNames from "classnames";
import { ModeEnum } from "../../typings/ColorPickerProps";

export interface ButtonProps {
    color: string | undefined;
    mode: ModeEnum;
    onClick: () => void;
    disabled: boolean;
}

export const Button = (props: ButtonProps): ReactElement => {
    const { mode, disabled, onClick, color } = props;
    return (
        <button
            className={classNames("btn", {
                "widget-color-picker-input": mode === "input",
                hidden: mode === "inline",
                disabled
            })}
            onClick={onClick}
        >
            <div
                className={mode === "input" ? "widget-color-picker-input-inner" : "widget-color-picker-inner"}
                style={{ background: color }}
            />
        </button>
    );
};
