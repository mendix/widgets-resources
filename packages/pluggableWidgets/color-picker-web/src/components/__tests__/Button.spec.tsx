import { createElement } from "react";
import { render } from "@testing-library/react";

import { shallow, ShallowWrapper } from "enzyme";

import { Button, ButtonProps } from "../Button";

describe("Button", () => {
    const renderButton = (props: ButtonProps): ShallowWrapper<ButtonProps, any> => shallow(<Button {...props} />);
    const buttonProps: ButtonProps = {
        color: "#000000",
        disabled: false,
        mode: "popover",
        onClick: jest.fn()
    };

    it("render DOM structure", () => {
        const { asFragment } = render(<Button {...buttonProps} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders the structure correctly", () => {
        const buttonComponent = renderButton(buttonProps);

        expect(buttonComponent.hasClass("btn")).toBe(true);

        buttonComponent.setProps({ disabled: true });
        expect(buttonComponent.hasClass("disabled")).toBe(true);

        buttonComponent.setProps({ mode: "input" });
        expect(buttonComponent.hasClass("widget-color-picker-input")).toBe(true);

        buttonComponent.setProps({ mode: "inline" });
        expect(buttonComponent.hasClass("hidden")).toBe(true);
    });
});
