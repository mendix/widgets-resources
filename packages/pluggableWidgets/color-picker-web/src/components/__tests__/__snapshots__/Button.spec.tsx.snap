// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button render DOM structure 1`] = `
<DocumentFragment>
  <button
    class="btn"
  >
    <div
      class="widget-color-picker-inner"
      style="background: rgb(0, 0, 0);"
    />
  </button>
</DocumentFragment>
`;
